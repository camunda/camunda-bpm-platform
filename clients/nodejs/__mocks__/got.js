/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { jest } from "@jest/globals";
// There is no alternative of `requireActual` for esm as of now
// See `jest.importActual` progress in https://github.com/facebook/jest/pull/10976.
// const got = jest.requireActual("got");
// const got = jest.createMockFromModule("got");

const handleRequest = (url, { testResult }) => {
  if (testResult instanceof Error) {
    return Promise.reject(testResult);
  }
  return {
    body:
      testResult instanceof Object
        ? JSON.stringify(testResult)
        : Buffer.from(testResult || "", "utf-8"),
    headers: {
      "content-type":
        testResult instanceof Object
          ? "application/json"
          : "application/octet-stream",
    },
    // headers: {},
  };
};

const gotMock = handleRequest;
gotMock.json = () => {};

const myModule = jest.fn().mockImplementation(gotMock);

class HTTPError extends Error {}
class RequestError extends Error {}
myModule.HTTPError = HTTPError;
myModule.RequestError = RequestError;

export default myModule;
export { myModule as got, HTTPError, RequestError };
