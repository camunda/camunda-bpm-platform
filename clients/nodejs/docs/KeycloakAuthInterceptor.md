# KeycloakAuthInterceptor

A KeycloakAuthInterceptor instance is an interceptor that adds a Bearer token header, containing 
a [Keycloak](https://www.keycloak.org/) access token, to all requests. This interceptor can be used
if the Camunda REST API is protected with [Keycloak Gatekeeper](https://github.com/keycloak/keycloak-gatekeeper).

This client serves also as an example for OpenID Connect based authentication. It shows also how an interceptor
with async functionality can be implemented with [hooks](https://github.com/sindresorhus/got#hooks), which are 
provided by the underlying [got](https://github.com/sindresorhus/got) HTTP request library. 

```js
const {
  Client,
  KeycloakAuthInterceptor
} = require("camunda-external-task-client-js");

const keycloakAuthentication = new KeycloakAuthInterceptor({
  tokenEndpoint: "https://your.keyclock.domain/realms/your-realm/protocol/openid-connect/token",
  clientId: "your-client-id",
  clientSecret: "your-client-secret"
});

const client = new Client({
  baseUrl: "http://localhost:8080/engine-rest",
  interceptors: keycloakAuthentication
});
```

## Caching

The Keycloak access token has an expiry defined. To reduce the requests to the token endpoint, we cache the token 
response as long it is valid.

To poll the API always with a valid token, we subtract the `cacheOffset` from the validity. If the token is 60 seconds
valid and we poll the API every 5 seconds, there could be the case that we poll the API exactly at the time the token 
expires. The default `cacheOffset` from 10 seconds is a good balance between the token validity and the average request 
duration.

## new KeycloakAuthInterceptor(options)

Here's a list of the available options:

| Option        | Description                           | Type   | Required | Default |
| ------------- | ------------------------------------- | ------ | -------- | ------- |
| tokenEndpoint | URL to the Keycloak token endpoint    | string | ✓        |         |
| clientId      | The Keycloak client id                | string | ✓        |         |
| clientSecret  | The Keycloak client secret            | string | ✓        |         |
| cacheOffset   | The time in seconds to subtract from the token expiry | number |          | 10      |
