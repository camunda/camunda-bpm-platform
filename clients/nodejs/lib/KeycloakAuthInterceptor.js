/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import got from "got";
import {
  MISSING_KEYCLOAK_AUTH_PARAMS,
  UNEXPECTED_KEYCLOAK_TOKEN_RESULT,
} from "./__internal/errors.js";

/**
 * A KeycloakAuthInterceptor instance is an interceptor that adds a Bearer token header, containing
 * a access token, to all requests. This interceptor can be used if the Camunda REST API is protected with
 * Keycloak Gatekeeper.
 */
class KeycloakAuthInterceptor {
  /**
   * The class constructor.
   *
   * @param options The Keycloak auth interceptor options.
   * @returns function(*): {hooks: {beforeRequest: [function(*): Promise<void>]}} The interceptor function.
   * @throws Error if the required options are not set.
   */
  constructor(options) {
    if (
      !options ||
      !options.tokenEndpoint ||
      !options.clientId ||
      !options.clientSecret
    ) {
      throw new Error(MISSING_KEYCLOAK_AUTH_PARAMS);
    }

    /**
     * Bind member methods
     */
    this.getAccessToken = this.getAccessToken.bind(this);
    this.cacheToken = this.cacheToken.bind(this);
    this.interceptor = this.interceptor.bind(this);

    this.cacheOffset =
      options.cacheOffset !== undefined ? options.cacheOffset : 10;
    this.tokenEndpoint = options.tokenEndpoint;
    this.clientId = options.clientId;
    this.clientSecret = options.clientSecret;

    return this.interceptor;
  }

  /**
   * Requests a new access token from the Keycloak endpoint.
   *
   * @param tokenEndpoint The URL to the Keycloak token endpoint.
   * @param clientID      The Keycloak client ID.
   * @param clientSecret  The Keycloak client secret.
   * @returns {Promise<{access_token: String, expires_in: Number}>} The token response, containing the access token and
   * it's expiry in seconds.
   * @throws Error if an error occurred during the request.
   */
  async getAccessToken(tokenEndpoint, clientID, clientSecret) {
    const credentials = Buffer.from(`${clientID}:${clientSecret}`).toString(
      "base64"
    );

    try {
      return await got(tokenEndpoint, {
        method: "POST",
        headers: {
          "Content-Type": "application/x-www-form-urlencoded",
          Authorization: `Basic: ${credentials}`,
        },
        body: "grant_type=client_credentials",
      }).json();
    } catch (e) {
      throw new Error(
        `${UNEXPECTED_KEYCLOAK_TOKEN_RESULT} status: ${e.response.statusCode}; body: ${e.response.body}`
      );
    }
  }

  /**
   * The Keycloak access token has an expiry defined. To reduce the requests to the token endpoint, we
   * cache the token response as long it is valid.
   *
   * To poll the API always with a valid token, we subtract the `cacheOffset` from the validity. If the token is 60
   * seconds valid and we poll the API every 5 seconds, there could be the case that we poll the API exactly at the
   * time the token expires. The default `cacheOffset` from 10 seconds is a good balance between the token validity
   * and the average request duration.
   *
   * If the `expires_in` property wasn't set or if the validity is less than or equal the `cacheOffset`, we don't
   * cache the token.
   *
   * @param {{access_token: String, expires_in: Number}} token The token response.
   */
  cacheToken(token) {
    const expiresIn = token.expires_in || 0;

    if (expiresIn > this.cacheOffset) {
      this.tokenCache = token;
      const tokenCleaner = () => {
        this.tokenCache = null;
      };
      setTimeout(
        tokenCleaner.bind(this),
        (expiresIn - this.cacheOffset) * 1000
      );
    }
  }

  /**
   * The interceptor function that enriches the `got` request with the access token hook.
   *
   * @param config The `got` request config.
   * @returns {{hooks: {beforeRequest: [function(*): Promise<void>]}}} The `got` config containing the access
   * token hook.
   */
  interceptor(config) {
    // https://www.npmjs.com/package/got#hooksbeforerequest
    const hooks = {
      beforeRequest: [
        async (options) => {
          let token = this.tokenCache;
          if (!token) {
            token = await this.getAccessToken(
              this.tokenEndpoint,
              this.clientId,
              this.clientSecret
            );
            this.cacheToken(token);
          }

          if (token && token.access_token) {
            const defaultHeaders = options.headers || {};
            const headers = {
              Authorization: `Bearer ${token.access_token}`,
            };

            options.headers = { ...defaultHeaders, ...headers };
          } else {
            throw new Error(
              `${UNEXPECTED_KEYCLOAK_TOKEN_RESULT} token without access_token property: ${JSON.stringify(
                token
              )}`
            );
          }
        },
      ],
    };

    return { ...config, hooks: { ...config.hooks, ...hooks } };
  }
}

export default KeycloakAuthInterceptor;
