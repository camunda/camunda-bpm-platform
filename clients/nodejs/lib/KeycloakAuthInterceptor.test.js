/*
 * Copyright Camunda Services GmbH and/or licensed to Camunda Services GmbH
 * under one or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information regarding copyright
 * ownership. Camunda licenses this file to you under the Apache License,
 * Version 2.0; you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { jest } from "@jest/globals";
import got from "got";
import KeycloakAuthInterceptor from "./KeycloakAuthInterceptor.js";
import {
  MISSING_KEYCLOAK_AUTH_PARAMS,
  UNEXPECTED_KEYCLOAK_TOKEN_RESULT,
} from "./__internal/errors.js";

describe("KeycloakAuthInterceptor", () => {
  afterEach(() => {
    got.mockClear();
  });

  test("should throw error if tokenEndpoint, clientId or clientSecret are missing", () => {
    expect(() => new KeycloakAuthInterceptor()).toThrowError(
      MISSING_KEYCLOAK_AUTH_PARAMS
    );
    expect(
      () => new KeycloakAuthInterceptor({ tokenEndpoint: "some endpoint" })
    ).toThrowError(MISSING_KEYCLOAK_AUTH_PARAMS);
    expect(
      () => new KeycloakAuthInterceptor({ clientId: "some id" })
    ).toThrowError(MISSING_KEYCLOAK_AUTH_PARAMS);
    expect(
      () => new KeycloakAuthInterceptor({ clientSecret: "some secret" })
    ).toThrowError(MISSING_KEYCLOAK_AUTH_PARAMS);
    expect(
      () =>
        new KeycloakAuthInterceptor({
          tokenEndpoint: "some endpoint",
          clientId: "some id",
        })
    ).toThrowError(MISSING_KEYCLOAK_AUTH_PARAMS);
    expect(
      () =>
        new KeycloakAuthInterceptor({
          tokenEndpoint: "some endpoint",
          clientSecret: "some secret",
        })
    ).toThrowError(MISSING_KEYCLOAK_AUTH_PARAMS);
    expect(
      () =>
        new KeycloakAuthInterceptor({
          clientId: "some password",
          clientSecret: "some secret",
        })
    ).toThrowError(MISSING_KEYCLOAK_AUTH_PARAMS);
  });

  test("should throw error if token endpoint returns unexpected HTTP response", async () => {
    // given
    const response = {
      body: "Some keycloak error",
      statusCode: 400,
      statusMessage: "Bad request",
    };
    const error = new got.HTTPError(response);
    error.response = response;
    got.mockReturnValue({ json: () => Promise.reject(error) });
    const keycloakAuthInterceptor = new KeycloakAuthInterceptor({
      tokenEndpoint: "some endpoint",
      clientId: "some id",
      clientSecret: "some secret",
    });

    // when
    const { hooks } = keycloakAuthInterceptor({});
    const hook = hooks.beforeRequest[0];

    // then
    try {
      await hook({});
    } catch (e) {
      expect(e).toEqual(
        new Error(
          `${UNEXPECTED_KEYCLOAK_TOKEN_RESULT} status: 400; body: Some keycloak error`
        )
      );
    }
  });

  test("should throw error if token doesn't contains the access token", async () => {
    // given
    got.mockReturnValue({ json: () => Promise.resolve({}) });
    const keycloakAuthInterceptor = new KeycloakAuthInterceptor({
      tokenEndpoint: "some endpoint",
      clientId: "some id",
      clientSecret: "some secret",
    });

    // when
    const { hooks } = keycloakAuthInterceptor({});
    const hook = hooks.beforeRequest[0];

    // then
    try {
      await hook({});
    } catch (e) {
      expect(e).toEqual(
        new Error(
          `${UNEXPECTED_KEYCLOAK_TOKEN_RESULT} token without access_token property: {}`
        )
      );
    }
  });

  test("should add auth token to intercepted config", async () => {
    // given
    got.mockReturnValue({
      json: () => Promise.resolve({ access_token: "1234567890" }),
    });
    const keycloakAuthInterceptor = new KeycloakAuthInterceptor({
      tokenEndpoint: "some endpoint",
      clientId: "some id",
      clientSecret: "some secret",
    });
    const options = { key: "some value" };

    // when
    const { hooks } = keycloakAuthInterceptor({});
    const hook = hooks.beforeRequest[0];
    await hook(options);

    // then
    expect(options).toMatchSnapshot();
  });

  test("should cache the token response if token expiry is greater than cacheOffset", async () => {
    // given
    const tokenResponse = { access_token: "1234567890", expires_in: 5 };
    got.mockReturnValue({ json: () => Promise.resolve(tokenResponse) });
    const keycloakAuthInterceptor = new KeycloakAuthInterceptor({
      tokenEndpoint: "some endpoint",
      clientId: "some id",
      clientSecret: "some secret",
      cacheOffset: 0,
    });

    // when
    const { hooks } = keycloakAuthInterceptor({});
    const hook = hooks.beforeRequest[0];
    await hook({});
    await hook({});

    // then
    expect(got).toHaveBeenCalledTimes(1);
  });

  test("should not cache the token response if token expiry is less than cacheOffset", async () => {
    // given
    const tokenResponse = { access_token: "1234567890", expires_in: 5 };
    got.mockReturnValue({ json: () => Promise.resolve(tokenResponse) });
    const keycloakAuthInterceptor = new KeycloakAuthInterceptor({
      tokenEndpoint: "some endpoint",
      clientId: "some id",
      clientSecret: "some secret",
      cacheOffset: 10,
    });

    // when
    const { hooks } = keycloakAuthInterceptor({});
    const hook = hooks.beforeRequest[0];
    await hook({});
    await hook({});

    // then
    expect(got).toHaveBeenCalledTimes(2);
  });

  test("should clear the token cache", async () => {
    jest.useFakeTimers();

    // given
    const tokenResponse = { access_token: "1234567890", expires_in: 5 };
    got.mockReturnValue({ json: () => Promise.resolve(tokenResponse) });
    const keycloakAuthInterceptor = new KeycloakAuthInterceptor({
      tokenEndpoint: "some endpoint",
      clientId: "some id",
      clientSecret: "some secret",
      cacheOffset: 0,
    });

    // when
    const { hooks } = keycloakAuthInterceptor({});
    const hook = hooks.beforeRequest[0];
    await hook({});
    jest.runAllTimers();
    await hook({});

    // then
    expect(got).toHaveBeenCalledTimes(2);

    jest.useRealTimers();
  });
});
