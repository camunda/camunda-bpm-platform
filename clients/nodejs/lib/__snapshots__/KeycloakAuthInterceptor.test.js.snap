// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeycloakAuthInterceptor should add auth token to intercepted config 1`] = `
{
  "headers": {
    "Authorization": "Bearer 1234567890",
  },
  "key": "some value",
}
`;
